pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/RollupCollector.java

```diff
@@ -0,0 +1,138 @@
+
+  private final int _keySize;
+  private final int _valueSize;
+  private final String[] _keyColumns;
+  private final String[] _valueColumns;
+  private final ValueAggregator[] _valueAggregators;
+  private final MetricFieldSpec[] _metricFieldSpecs;
+
+  public RollupCollector(CollectorConfig collectorConfig, Schema schema) {
+    _keySize = schema.getColumnNames().size() - schema.getMetricNames().size();
+    _valueSize = schema.getMetricNames().size();
+    _keyColumns = new String[_keySize];
+    _valueColumns = new String[_valueSize];
+    _valueAggregators = new ValueAggregator[_valueSize];
+    _metricFieldSpecs = new MetricFieldSpec[_valueSize];
+
+    Map<String, ValueAggregatorFactory.ValueAggregatorType> aggregatorTypeMap = collectorConfig.getAggregatorTypeMap();
+    if (aggregatorTypeMap == null) {
+      aggregatorTypeMap = Collections.emptyMap();
+    }
+    int valIdx = 0;
+    int keyIdx = 0;
+    for (FieldSpec fieldSpec : schema.getAllFieldSpecs()) {
+      if (!fieldSpec.isVirtualColumn()) {
+        String name = fieldSpec.getName();
+        if (fieldSpec.getFieldType().equals(FieldSpec.FieldType.METRIC)) {
+          _metricFieldSpecs[valIdx] = (MetricFieldSpec) fieldSpec;
+          _valueColumns[valIdx] = name;
+          _valueAggregators[valIdx] = ValueAggregatorFactory.getValueAggregator(
+              aggregatorTypeMap.getOrDefault(name, ValueAggregatorFactory.ValueAggregatorType.SUM).toString());
+          valIdx++;
+        } else {
+          _keyColumns[keyIdx++] = name;
+        }
+      }
+    }
+  }
+
+  /**
+   * If a row already exists in the collection (based on dimension + time columns), rollup the metric values, else add the row
+   */
+  @Override
+  public void collect(GenericRow genericRow) {
+    Object[] key = new Object[_keySize];
+    for (int i = 0; i < _keySize; i++) {
+      key[i] = genericRow.getValue(_keyColumns[i]);
+    }
+    Record keyRecord = new Record(key);
+    GenericRow prev = _collection.get(keyRecord);
+    if (prev == null) {
+      _collection.put(keyRecord, genericRow);
+    } else {
+      for (int i = 0; i < _valueSize; i++) {
+        String valueColumn = _valueColumns[i];
+        Object aggregate = _valueAggregators[i]
+            .aggregate(prev.getValue(valueColumn), genericRow.getValue(valueColumn), _metricFieldSpecs[i]);
+        prev.putValue(valueColumn, aggregate);
+      }
+    }
+  }
+
+  @Override
+  public Iterator<GenericRow> iterator() {
+    return _collection.values().iterator();
+  }
+
+  @Override
+  public int size() {
+    return _collection.size();
+  }
+
+  @Override
+  public void reset() {
+    _collection.clear();
+  }
+
+  /**
+   * A record representation for the keys of the record
+   * Note that the dimensions can have multi-value columns, and hence the equals and hashCode need deep array operations
+   */
+  private static class Record {
+    private final Object[] _values;
```

Comment:
Are these the key parts? If so, can we name the member as `_keyParts`? `_values` gets confusing.